core/opregion.c

```diff
@@ -38,571 +41,327 @@ void lai_write_opregion(lai_nsnode_t *field, lai_variable_t *source) {
-        }
-    } else if (opregion->op_address_space == ACPI_OPREGION_IO) {
-        // I/O port
-        switch (field->fld_flags & 0x0F) {
-            case FIELD_BYTE_ACCESS:
-                if (!laihost_inb)
-                    lai_panic("host does not provide port I/O functions");
-                value = (uint64_t)laihost_inb(opregion->op_base + offset);
-                break;
-            case FIELD_WORD_ACCESS:
-                if (!laihost_inw)
-                    lai_panic("host does not provide port I/O functions");
-                value = (uint64_t)laihost_inw(opregion->op_base + offset);
-                break;
-            case FIELD_DWORD_ACCESS:
-            case FIELD_ANY_ACCESS:
-                if (!laihost_ind)
-                    lai_panic("host does not provide port I/O functions");
-                value = (uint64_t)laihost_ind(opregion->op_base + offset);
-                break;
-            default:
-                lai_panic("undefined field flags 0x%02X: %s", field->fld_flags,
-                          lai_stringify_node_path(field));
-        }
-    } else if (opregion->op_address_space == ACPI_OPREGION_MEMORY) {
-        // Memory-mapped I/O
-        if (!laihost_map)
-            lai_panic("host does not provide memory mapping functions");
-        mmio = laihost_map(opregion->op_base + offset, 8);
-        uint8_t *mmio_byte;
-        uint16_t *mmio_word;
-        uint32_t *mmio_dword;
-        uint64_t *mmio_qword;
-
-        switch (field->fld_flags & 0x0F) {
-            case FIELD_BYTE_ACCESS:
-                mmio_byte = (uint8_t *)mmio;
-                value = (uint64_t)mmio_byte[0];
-                break;
-            case FIELD_WORD_ACCESS:
-                mmio_word = (uint16_t *)mmio;
-                value = (uint64_t)mmio_word[0];
-                break;
-            case FIELD_DWORD_ACCESS:
-            case FIELD_ANY_ACCESS:
-                mmio_dword = (uint32_t *)mmio;
-                value = (uint64_t)mmio_dword[0];
-                break;
-            case FIELD_QWORD_ACCESS:
-                mmio_qword = (uint64_t *)mmio;
-                value = mmio_qword[0];
-                break;
-            default:
-                lai_panic("undefined field flags 0x%02X: %s", field->fld_flags,
-                          lai_stringify_node_path(field));
-        }
-    } else if (opregion->op_address_space == ACPI_OPREGION_PCI) {
-        LAI_CLEANUP_STATE lai_state_t state;
-        lai_init_state(&state);
-
-        // PCI seg number is in the _SEG object.
-        lai_nsnode_t *seg_handle = lai_resolve_search(opregion, "_SEG");
-        if (seg_handle) {
-            if (lai_eval(&seg_number, seg_handle, &state))
-                lai_panic("could not evaluate _SEG of OperationRegion()");
-            bbn_result = seg_number.integer;
-        }
+    uint64_t read_offset = (field->fld_offset >> access_shift) * access_size / 8;
+    uint64_t bit_offset = field->fld_offset & (access_size - 1);
```

Comment:
I would move `bit_offset` into the loop and redefine it as `uint64_t bit_offset = (field->fld_offset + progress) & (access_size - 1);` since that is the only expression it is ever used in.